keps/sig-apps/4017-pod-index-label/README.md

```diff
@@ -0,0 +1,896 @@
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+Currently, StatefulSet pods do not have their pod index as a label or annotation, and
+Indexed Jobs only set the pod completion index as an annotation.
+This KEP proposes to set the pod index as a label on pods for StatefulSets and Indexed Jobs.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+StatefulSets, similar to Indexed Jobs, assign a pod ordinal to each of their pods (this is
+analogous to the pod completion index for Indexed Jobs).
+
+However, unlike Indexed Jobs, StatefulSet pods only have this ordinal as part of the pod name;
+they do not have it set as a annotation. This means that the only way for the StatefulSet pod to know its own index is
+to [parse the pod name](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/statefulset/stateful_set_utils.go#L63-L75),
+which is somewhat hacky solution. Furthermore, the pod index cannot be utilized by the downward API, 
+nor be used for selecting pods based on their index or filtering metrics/logs.
+
+Indexed Jobs, on the other hand, *do* have the pod completion index set as an annotation, which can
+be used by the downard API. However, annotations cannot be easily used by selectors, or for filtering metrics/logs.
+
+To address these issues, we propose to set the pod index as a label on pods for StatefulSets and
+Indexed Jobs.
+
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+- Set pod index as label on pods for StatefulSet/Indexed Jobs.
+- Adding the label can't be disruptive to existing workloads.
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+At a high level, the proposal is to modify the StatefulSet and Job controllers to set the pod index/ordinal
+as a label on the pod at pod creation time. For Jobs, this would only apply to jobs in
+Indexed completion mode. The details of this are outlined in the Design Details section below.
+
+StatefulSet pod label: `statefulset.kubernetes.io/pod-ordinal`
```

Comment:
This may be worth asking in sig-apps as I remember there was some discussion about this in a previous meeting.  

https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/apps/v1beta1/types.go#L29

Looks to use `statefulset.kubernetes.io` but we used `batch.kubernetes.io` in batch as there was already an existing label.

Reply:
Thanks, then we should continue to use `statefulset.kubernetes.io`

Reply:
makes sense, done.